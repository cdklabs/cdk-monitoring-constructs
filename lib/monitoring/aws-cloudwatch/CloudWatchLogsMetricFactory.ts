import { DimensionsMap } from "aws-cdk-lib/aws-cloudwatch";

import {
  BaseMetricFactory,
  BaseMetricFactoryProps,
  MetricFactory,
  MetricStatistic,
} from "../../common";

const CloudWatchLogsNamespace = "AWS/Logs";

export interface CloudWatchLogsMetricFactoryProps
  extends BaseMetricFactoryProps {
  /**
   * Name of the log group to monitor.
   */
  readonly logGroupName: string;
}

export class CloudWatchLogsMetricFactory extends BaseMetricFactory<CloudWatchLogsMetricFactoryProps> {
  private readonly dimensionsMap: DimensionsMap;

  constructor(
    metricFactory: MetricFactory,
    props: CloudWatchLogsMetricFactoryProps,
  ) {
    super(metricFactory, props);

    this.dimensionsMap = {
      LogGroupName: props.logGroupName,
    };
  }

  metricIncomingLogEvents() {
    return this.metricFactory.createMetric(
      "IncomingLogEvents",
      MetricStatistic.N,
      "Logs",
      this.dimensionsMap,
      undefined,
      CloudWatchLogsNamespace,
      undefined,
      this.region,
      this.account,
    );
  }
}
