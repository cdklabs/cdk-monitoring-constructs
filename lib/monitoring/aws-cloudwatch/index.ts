export * from "./CloudWatchLogsMetricFactory";
export * from "./LogMonitoring";
