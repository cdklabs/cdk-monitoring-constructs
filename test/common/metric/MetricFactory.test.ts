import { Duration } from "aws-cdk-lib";
import { Metric } from "aws-cdk-lib/aws-cloudwatch";

import {
  MetricFactory,
  MetricFactoryDefaults,
  MetricStatistic,
  RateComputationMethod,
} from "../../../lib";

const DummyColor = "#abcdef";

test("createMetric without global namespace throws an error", () => {
  const metricFactory = new MetricFactory();

  expect(() =>
    metricFactory.createMetric(
      "DummyMetricName-NoOptionalParams",
      MetricStatistic.P90
    )
  ).toThrowError();
});

describe("snapshot test: global defaults", () => {
  function testMetric(props: MetricFactoryDefaults) {
    const metricFactory = new MetricFactory({
      globalDefaults: {
        namespace: "DummyNamespace",
        ...props,
      },
    });

    const metric = metricFactory.createMetric(
      "DummyMetricName",
      MetricStatistic.P90
    );
    const metricMath = metricFactory.createMetricMath(
      "DummyExpression",
      {},
      "label"
    );
    const metricSearch = metricFactory.createMetricSearch(
      "DummyQuery",
      {},
      MetricStatistic.P90
    );
    const metricAnomalyDetection = metricFactory.createMetricAnomalyDetection(
      new Metric({
        metricName: "DummyMetric1",
        namespace: "DummyNamespace",
      }),
      2,
      "label"
    );

    expect({
      metric,
      metricMath,
      metricSearch,
      metricAnomalyDetection,
    }).toMatchSnapshot();
  }

  test("with namespace", () => {
    testMetric({
      namespace: "CustomTestNamespace",
    });
  });

  test("with period", () => {
    testMetric({
      period: Duration.minutes(15),
    });
  });

  test("with region", () => {
    testMetric({
      region: "us-west-2",
    });
  });

  test("with account", () => {
    testMetric({
      account: "123456789",
    });
  });

  test("with multiple props", () => {
    testMetric({
      namespace: "CustomTestNamespace",
      period: Duration.minutes(15),
      region: "us-west-2",
      account: "123456789",
    });
  });
});

test("snapshot test: createMetric", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const metricWithNoOptionalParams = metricFactory.createMetric(
    "DummyMetricName-NoOptionalParams",
    MetricStatistic.P90
  );

  expect(metricWithNoOptionalParams).toMatchSnapshot();

  const metricWithAllOptionalParams = metricFactory.createMetric(
    "DummyMetricName-AllOptionalParams",
    MetricStatistic.P90,
    "DummyLabel",
    { DummyDimension: "DummyDimensionValue" },
    DummyColor,
    "DummyNamespaceOverride"
  );

  expect(metricWithAllOptionalParams).toMatchSnapshot();

  const metricWithUndefinedDimensions = metricFactory.createMetric(
    "DummyMetricName-AllOptionalParams",
    MetricStatistic.P90,
    "DummyLabel",
    {
      DummyDimension: undefined as unknown as string,
      AnotherDummyDimension: undefined as unknown as string,
    },
    DummyColor,
    "DummyNamespaceOverride",
    Duration.minutes(15),
    "us-west-2",
    "123456789"
  );

  expect(metricWithUndefinedDimensions).toMatchSnapshot();
});

test("snapshot test: createMetricMath", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const a = metricFactory.createMetric("a", MetricStatistic.SUM);
  const b = metricFactory.createMetric("b", MetricStatistic.SUM);

  const metricWithNoOptionalParams = metricFactory.createMetricMath(
    "a+b",
    { a, b },
    "a and b"
  );

  expect(metricWithNoOptionalParams).toMatchSnapshot();

  const metricWithAllOptionalParams = metricFactory.createMetricMath(
    "a+b",
    { a, b },
    "a and b",
    DummyColor
  );

  expect(metricWithAllOptionalParams).toMatchSnapshot();
});

test("snapshot test: toRate with detail", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const metric = metricFactory.createMetric(
    "Metric",
    MetricStatistic.SUM,
    "Label"
  );

  const metricAverage = metricFactory.toRate(
    metric,
    RateComputationMethod.AVERAGE,
    true
  );
  expect(metricAverage).toMatchSnapshot();

  const metricPerSecond = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_SECOND,
    true
  );
  expect(metricPerSecond).toMatchSnapshot();

  const metricPerMinute = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_MINUTE,
    true
  );
  expect(metricPerMinute).toMatchSnapshot();

  const metricPerHour = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_HOUR,
    true
  );
  expect(metricPerHour).toMatchSnapshot();

  const metricPerDay = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_DAY,
    true
  );
  expect(metricPerDay).toMatchSnapshot();
});

test("snapshot test: toRate without detail", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const metric = metricFactory.createMetric(
    "Metric",
    MetricStatistic.SUM,
    "Label"
  );

  const metricAverage = metricFactory.toRate(
    metric,
    RateComputationMethod.AVERAGE,
    false
  );
  expect(metricAverage).toMatchSnapshot();

  const metricPerSecond = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_SECOND,
    false
  );
  expect(metricPerSecond).toMatchSnapshot();

  const metricPerMinute = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_MINUTE,
    false
  );
  expect(metricPerMinute).toMatchSnapshot();

  const metricPerHour = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_HOUR,
    false
  );
  expect(metricPerHour).toMatchSnapshot();

  const metricPerDay = metricFactory.toRate(
    metric,
    RateComputationMethod.PER_DAY,
    false
  );
  expect(metricPerDay).toMatchSnapshot();
});

test("snapshot test: createMetricSearch", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const metricWithNoOptionalParams = metricFactory.createMetricSearch(
    "MyMetricPrefix-",
    { DummyDimension: "DummyDimensionValue" },
    MetricStatistic.SUM
  );

  expect(metricWithNoOptionalParams).toMatchSnapshot();

  const metricWithAllOptionalParams = metricFactory.createMetricSearch(
    "MyMetricPrefix-",
    { DummyDimension: "DummyDimensionValue" },
    MetricStatistic.SUM,
    "DummyNamespaceOverride"
  );

  expect(metricWithAllOptionalParams).toMatchSnapshot();

  const metricWithUndefinedDimensions = metricFactory.createMetricSearch(
    "MyMetricPrefix-",
    {
      DummyUndefinedDimension: undefined as unknown as string,
      DummyDefinedDimension: "DummyDimensionValue",
    },
    MetricStatistic.SUM,
    "DummyNamespaceOverride"
  );

  expect(metricWithUndefinedDimensions).toMatchSnapshot();

  const metricWithEmptyDimensions = metricFactory.createMetricSearch(
    "MyMetricPrefix-",
    {},
    MetricStatistic.SUM,
    "DummyNamespaceOverride"
  );

  expect(metricWithEmptyDimensions).toMatchSnapshot();
});

test("snapshot test: createMetricAnomalyDetection", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const metric = new Metric({
    namespace: "DummyNamespace",
    metricName: "DummyMetric",
  });

  const anomaly = metricFactory.createMetricAnomalyDetection(
    metric,
    2,
    "DummyLabel"
  );

  expect(anomaly).toMatchSnapshot();
});

test("multiply/divide metric produces expected result", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  const metric = new Metric({
    namespace: "DummyNamespace",
    metricName: "DummyMetric",
  });

  expect(() =>
    metricFactory.multiplyMetric(metric, 0.5, "DummyLabel")
  ).toThrowError();
  expect(() =>
    metricFactory.divideMetric(metric, 0.6, "DummyLabel")
  ).toThrowError();

  expect(metricFactory.multiplyMetric(metric, 1, "DummyLabel")).toStrictEqual(
    metric
  );
  expect(metricFactory.divideMetric(metric, 1, "DummyLabel")).toStrictEqual(
    metric
  );

  expect(
    metricFactory.multiplyMetric(metric, 42, "DummyLabel")
  ).toMatchSnapshot();
  expect(metricFactory.divideMetric(metric, 5, "DummyLabel")).toMatchSnapshot();
});

test("sanitizeMetricExpressionIdSuffix removes unwanted characters", () => {
  const metricFactory = new MetricFactory({
    globalDefaults: {
      namespace: "DummyNamespace",
    },
  });

  expect(
    metricFactory.sanitizeMetricExpressionIdSuffix("Name With Spaces")
  ).toStrictEqual("NameWithSpaces");
  expect(
    metricFactory.sanitizeMetricExpressionIdSuffix("-Hyphen_Underscore")
  ).toStrictEqual("Hyphen_Underscore");
  expect(
    metricFactory.sanitizeMetricExpressionIdSuffix("Alpha1234567890Numeric")
  ).toStrictEqual("Alpha1234567890Numeric");
  expect(
    metricFactory.sanitizeMetricExpressionIdSuffix(
      "Some Weird Characters: +ěščřžýáíé='"
    )
  ).toStrictEqual("SomeWeirdCharacters");
});
