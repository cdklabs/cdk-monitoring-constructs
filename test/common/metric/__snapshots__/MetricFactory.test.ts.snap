// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiply/divide metric produces expected result 1`] = `
MathExpression {
  "color": undefined,
  "expression": "m1 * 42",
  "label": "DummyLabel",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "DummyMetric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Average",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`multiply/divide metric produces expected result 2`] = `
MathExpression {
  "color": undefined,
  "expression": "m1 / 5",
  "label": "DummyLabel",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "DummyMetric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Average",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: createMetric 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": undefined,
  "label": undefined,
  "metricName": "DummyMetricName-NoOptionalParams",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: createMetric 2`] = `
Metric {
  "account": undefined,
  "color": "#abcdef",
  "dimensions": Object {
    "DummyDimension": "DummyDimensionValue",
  },
  "label": "DummyLabel",
  "metricName": "DummyMetricName-AllOptionalParams",
  "namespace": "DummyNamespaceOverride",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: createMetric 3`] = `
Metric {
  "account": "123456789",
  "color": "#abcdef",
  "dimensions": Object {},
  "label": "DummyLabel",
  "metricName": "DummyMetricName-AllOptionalParams",
  "namespace": "DummyNamespaceOverride",
  "period": Duration {
    "amount": 15,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": "us-west-2",
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: createMetricAnomalyDetection 1`] = `
AnomalyDetectionMathExpression {
  "color": undefined,
  "expression": "ANOMALY_DETECTION_BAND(m1,2)",
  "label": "DummyLabel",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "DummyMetric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Average",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: createMetricMath 1`] = `
MathExpression {
  "color": undefined,
  "expression": "a+b",
  "label": "a and b",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "a": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "a",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
    "b": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "b",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: createMetricMath 2`] = `
MathExpression {
  "color": "#abcdef",
  "expression": "a+b",
  "label": "a and b",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "a": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "a",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
    "b": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": undefined,
      "metricName": "b",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: createMetricSearch 1`] = `
MathExpression {
  "color": undefined,
  "expression": "SEARCH('{DummyNamespace,DummyDimension} DummyDimension=\\"DummyDimensionValue\\" MyMetricPrefix-', 'Sum', 300)",
  "label": " ",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {},
  "warnings": Array [
    "Math expression 'SEARCH('{DummyNamespace,DummyDimension} DummyDimension=\\"DummyDimensionValue\\" MyMetricPrefix-', 'Sum', 300)' references unknown identifiers: ummyNamespace, ummyDimension, ummyDimension, ummyDimensionValue, yMetricPrefix, um. Please add them to the 'usingMetrics' map.",
  ],
}
`;

exports[`snapshot test: createMetricSearch 2`] = `
MathExpression {
  "color": undefined,
  "expression": "SEARCH('{DummyNamespaceOverride,DummyDimension} DummyDimension=\\"DummyDimensionValue\\" MyMetricPrefix-', 'Sum', 300)",
  "label": " ",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {},
  "warnings": Array [
    "Math expression 'SEARCH('{DummyNamespaceOverride,DummyDimension} DummyDimension=\\"DummyDimensionValue\\" MyMetricPrefix-', 'Sum', 300)' references unknown identifiers: ummyNamespaceOverride, ummyDimension, ummyDimension, ummyDimensionValue, yMetricPrefix, um. Please add them to the 'usingMetrics' map.",
  ],
}
`;

exports[`snapshot test: createMetricSearch 3`] = `
MathExpression {
  "color": undefined,
  "expression": "SEARCH('{DummyNamespaceOverride,DummyUndefinedDimension,DummyDefinedDimension} DummyDefinedDimension=\\"DummyDimensionValue\\" MyMetricPrefix-', 'Sum', 300)",
  "label": " ",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {},
  "warnings": Array [
    "Math expression 'SEARCH('{DummyNamespaceOverride,DummyUndefinedDimension,DummyDefinedDimension} DummyDefinedDimension=\\"DummyDimensionValue\\" MyMetricPrefix-', 'Sum', 300)' references unknown identifiers: ummyNamespaceOverride, ummyUndefinedDimension, ummyDefinedDimension, ummyDefinedDimension, ummyDimensionValue, yMetricPrefix, um. Please add them to the 'usingMetrics' map.",
  ],
}
`;

exports[`snapshot test: createMetricSearch 4`] = `
MathExpression {
  "color": undefined,
  "expression": "SEARCH('{DummyNamespaceOverride}  MyMetricPrefix-', 'Sum', 300)",
  "label": " ",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {},
  "warnings": Array [
    "Math expression 'SEARCH('{DummyNamespaceOverride}  MyMetricPrefix-', 'Sum', 300)' references unknown identifiers: ummyNamespaceOverride, yMetricPrefix, um. Please add them to the 'usingMetrics' map.",
  ],
}
`;

exports[`snapshot test: global defaults 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": undefined,
  "label": undefined,
  "metricName": "DummyMetricName",
  "namespace": "SomeNamespace",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: global defaults 2`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": undefined,
  "label": undefined,
  "metricName": "DummyMetricName",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 15,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: global defaults 3`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": undefined,
  "label": undefined,
  "metricName": "DummyMetricName",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": "us-west-2",
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: global defaults 4`] = `
Metric {
  "account": "123456789",
  "color": undefined,
  "dimensions": undefined,
  "label": undefined,
  "metricName": "DummyMetricName",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: global defaults 5`] = `
Metric {
  "account": "123456789",
  "color": undefined,
  "dimensions": undefined,
  "label": undefined,
  "metricName": "DummyMetricName",
  "namespace": "SomeNamespace",
  "period": Duration {
    "amount": 15,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": "us-west-2",
  "statistic": "p90",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: toRate with detail 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": undefined,
  "label": "Label (avg) (min: \${MIN}, max: \${MAX})",
  "metricName": "Metric",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "Average",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: toRate with detail 2`] = `
MathExpression {
  "color": undefined,
  "expression": "m1 / PERIOD(m1)",
  "label": "Label/s (min: \${MIN}, max: \${MAX}, avg: \${AVG})",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate with detail 3`] = `
MathExpression {
  "color": undefined,
  "expression": "(60 * m1) / PERIOD(m1)",
  "label": "Label/m (min: \${MIN}, max: \${MAX}, avg: \${AVG})",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate with detail 4`] = `
MathExpression {
  "color": undefined,
  "expression": "(3600 * m1) / PERIOD(m1)",
  "label": "Label/h (min: \${MIN}, max: \${MAX}, avg: \${AVG})",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate with detail 5`] = `
MathExpression {
  "color": undefined,
  "expression": "(86400 * m1) / PERIOD(m1)",
  "label": "Label/d (min: \${MIN}, max: \${MAX}, avg: \${AVG})",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate without detail 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": undefined,
  "label": "Label (avg)",
  "metricName": "Metric",
  "namespace": "DummyNamespace",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "Average",
  "unit": undefined,
  "warnings": undefined,
}
`;

exports[`snapshot test: toRate without detail 2`] = `
MathExpression {
  "color": undefined,
  "expression": "m1 / PERIOD(m1)",
  "label": "Label/s",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate without detail 3`] = `
MathExpression {
  "color": undefined,
  "expression": "(60 * m1) / PERIOD(m1)",
  "label": "Label/m",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate without detail 4`] = `
MathExpression {
  "color": undefined,
  "expression": "(3600 * m1) / PERIOD(m1)",
  "label": "Label/h",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;

exports[`snapshot test: toRate without detail 5`] = `
MathExpression {
  "color": undefined,
  "expression": "(86400 * m1) / PERIOD(m1)",
  "label": "Label/d",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "searchAccount": undefined,
  "searchRegion": undefined,
  "usingMetrics": Object {
    "m1": Metric {
      "account": undefined,
      "color": undefined,
      "dimensions": undefined,
      "label": "Label",
      "metricName": "Metric",
      "namespace": "DummyNamespace",
      "period": Duration {
        "amount": 5,
        "unit": TimeUnit {
          "inMillis": 60000,
          "isoLabel": "M",
          "label": "minutes",
        },
      },
      "region": undefined,
      "statistic": "Sum",
      "unit": undefined,
      "warnings": undefined,
    },
  },
}
`;
